content/extending-chat-widget/custom-identity-provider/index.mdx

```diff
@@ -0,0 +1,180 @@
+
+```js
+    // Custom Identity Provider
+    window.__lc.custom_identity_provider = function() {
+      return {
+          getToken: ...,
+          getFreshToken: ...,
+          hasToken: ...,
+          invalidate: ...
+      }
+    }
+```
+
+# Methods
+
+In order to make it work, you'll have to properly implement and provide a set of following methods:
+
+- `getToken` - resolving [Chat Widget token](#chat-widget-token). If you want to cache the token somehow, this should return the cached token instead of a fresh request to API.
+- `getFreshToken` - resolving [Chat Widget token](#chat-widget-token). This should always make a call for a fresh token from accounts API.
+- `hasToken` - resolving boolean. Determining whether a token has been acquired.
+- `invalidate` - resolving nothing. When called should remove the current token. There is no need to do anything else, as a new token will be requested by `getFreshToken` afterward.
+
+All of them will have to be returned in a function assigned to `window.__lc.custom_identity_provider`.
+
+# Chat Widget Token
+
+LiveChat authorization token you are willing to resolve in functions should contain the following properties:
+
+| Parameter      | Description                                                                      |
+| -------------- | -------------------------------------------------------------------------------- |
+| `accessToken`  | A token you can use to call LiveChat APIs on behalf of the user (customer).      |
+| `entityId`     | The ID of the customer.                                                          |
+| `expiresIn`    | A number in **milliseconds** specifying how long the `accessToken` will be valid.|
+| `tokenType`    | Value: `Bearer`.                                                                 |
+| `creationDate` | Unix timestamp specyfing the time of creation of the token.                       |
+| `licenseId`    | LiveChat license ID.                                                             |
+
+## How to acquire it?
+
+In order to get the token, you need to implement a function that follows authorization flow by making a call to a `/customer` endpoint from LiveChat Accounts API. To do so, you can follow [creating a new customer flow](https://developers.livechat.com/docs/authorization/authorizing-api-calls/#creating-a-new-customer) or have a look at our [example](https://github.com/livechat/identity-provider-example/blob/master/lib/get-customer-token.ts) using cookies. Mind the fact that the [data](https://developers.livechat.com/docs/authorization/authorizing-api-calls/#response-4) returned from this endpoint is not 1:1 with the token shape described above. You will have to add the `creationDate` and `licenseId` fields. Additionally, we expect `expiresIn` to be defined in milliseconds while the endpoint operates in seconds. Therefore you should multiply the provided value by `1000` before passing it further.
```

Comment:
```suggestion
In order to get a Chat Widget token, you need to implement a function that calls the `/customer` endpoint. To do so, you can [create a new customer](https://developers.livechat.com/docs/authorization/authorizing-api-calls/#creating-a-new-customer) or have a look at our [example](https://github.com/livechat/identity-provider-example/blob/master/lib/get-customer-token.ts) that uses cookies. Mind the fact that the [data](https://developers.livechat.com/docs/authorization/authorizing-api-calls/#response-4) returned from the `enter-endpoint-here` endpoint is not 1:1 with the token shape described above. You'll have to make the following adjustments to your token before passing it further:

- Add the `creationDate` and `licenseId` fields. 
- Multiply `expiresIn` by 1000. It's because we expect `expiresIn` to be expressed in milliseconds while the `enter-endpoint-here` endpoint operates in seconds.
```

Reply:
A few comments:
1. Can we cut out "follow authorizaton flow by making a call"? It seems to be unnecessarily lengthy.
2. The LiveChat Accounts API is another name to Global Accounts API. I would avoid using "LiveChat Accounts" and just paste full endpoints to avoid confusion.

Reply:
`you can [create a new customer]` is not what we want to achieve, essentially we would like to just get token of rather exisiting customer which is defined in the docs under the link - if we do have entity_id. Its all really tricky to describe without proper documentation of this use-case on the accounts side

Reply:
probably it could be described a bit better but with following the flow, I just wanted the reader to study whats under this link and realise that they should have entity_id therefore its not really creating new customer but getting identity of already existing one. Still if they don't have a pair between their customer and ours (because it does not exist yet) they will have to create it so the name is valid only for the first execution of this logic

Reply:
I've tried to describe this better in recent commit

Reply:
Please, tak a look at this PR: https://github.com/livechat/livechat-public-docs/pull/964
Those are changes for Customer SSO, which will be released today. We got rid of the misleading "create a customer". You can link to the docs as it will be on prod before we merge your PR.

Reply:
I like the current description, though :) One remark: in the updates version of Customer SSO (customer auth docs) we call it "customer identity" not "new customer".